Assignment3/Q1_test.py

```diff
@@ -0,0 +1,49 @@
+"""Tests for question 1 - word search."""
+
+import unittest
+
+import Q1
+
+
+class Q1Test(unittest.TestCase):
```

Comment:
Optional suggestion about testing and constants.

The current tests require a bit of work to figure out what exactly is being tested because there are a number of variables being set in the `setUp()` method and then used in tests. The main concern is that if you look at one of the tests:
```python
  for word in self._given_desired_output:
    self.assertTrue(word in Q1.word_search(self._given_grid, self._given_dictionary))
```
it is not immediately obvious what is going on.

Consider, for comparison:
```python
  self.assertIn('CAT', Q1.word_search(_grid("CAD|TGO"), _dictionary('cat', 'dog'))
```
But how can we get this to work?

We can start by defining helper functions `_grid` and `_dictionary`:
```python
def _grid(string):
  """Creates a 2D array of characters from a string.

  The string represents the grid by having the characters in each row separated by a '|'.

  Args:
    string: a string, representation of the grid as a single string.

  Returns:
    a 2-D array of characters corresponding to the grid given.
  """
  grid = []
  for row in string.split('|'):
    grid.append([ch for ch in row])
    # Make sure each row has the same length.
    if len(grid[0]) != len(grid[-1]):
      raise ValueError("not all rows have the same length")
  return grid

def _dictionary(*words):
  """Create a dictionary with the given words.

  Args:
    *words: a list of strings, the words to be added to the dictionary.

  Returns:
    an instance of Q1.Dictionary containing the given words.
  """
  return Q1.Dictionary(words)
```
with these as helper functions defined in the test module, we can make the tests themselves more compact, without having to copy and paste too much.

It is sometimes helpful to have the `setUp()` create things for us, but that's mostly the case when things are shared by all the tests, not just some of them, like in this case.